src/main/python/community-detector/community_detector.py

```diff
@@ -23,41 +19,20 @@ def build_matrix(id_to_buckets):
     if len(id_to_buckets) == 0:
         return csr_matrix((0, 0), dtype=numpy.uint8)
 
-    data = numpy.ones(sum(map(len, id_to_buckets)), dtype=numpy.uint8)
+    max_el_id = max((item[0] for item in id_to_buckets))
+    data = numpy.ones(
+        sum((len(item[1]) for item in id_to_buckets)), dtype=numpy.uint8)
     indices = numpy.zeros(len(data), dtype=numpy.uint32)
-    indptr = numpy.zeros(len(id_to_buckets) + 1, dtype=numpy.uint32)
+    indptr = numpy.zeros(max_el_id + 2, dtype=numpy.uint32)
     pos = 0
-    for i, element in enumerate(id_to_buckets):
-        indices[pos:(pos + len(element))] = element
-        pos += len(element)
-        indptr[i + 1] = pos
+    for el_id, bucket in id_to_buckets:
+        indices[pos:(pos + len(bucket))] = bucket
+        pos += len(bucket)
+        indptr[el_id + 1:] = pos
```

Comment:
Maybe there is a performance gain we can squeeze here, avoiding to write always up to the end of the array.
Something like this (did not test this, please double check)
```python
    prev_el_id = 0
    prev_pos = 0

    for el_id, bucket in id_to_buckets:
        indices[pos:(pos + len(bucket))] = bucket
        pos += len(bucket)
        indptr[prev_el_id+2:el_id] = prev_pos

        prev_el_id = el_id
        prev_pos = post
    
    indptr[prev_el_id+1:] = prev_pos
```

Reply:
thanks! It's a good idea. I re-wrote it a little bit different and added one more test to be sure it works correctly for an edge case.